### PR TITLE
Static UnpatchAll and UnpatchID

### PR DESCRIPTION
**Pull request by ErisApps**
As a continuation following the discussion in PR #40, this PR handles the introduction of 2 (arguably) new static methods `Harmony.UnpatchAll()` and `Harmony.UnpatchID(string harmonyID)`.
Other than that, it also marks the old non-static `.UnpatchAll(string harmonyID = null)` as Obsolete.

Not sure how I can verify whether this actually works as intended, but the unit tests seem to run just fine, aside from a seemingly unrelated unit test that was already failing before the changes (`TestTraverse_Types.Traverse_InnerStatic`)

I think this covers everything that was mentioned in the other PR, but if I missed anything, please let me know.

### CODE REVIEW

File: Harmony/Public/Harmony.cs [hunk 1]
```diff
@@ -210,52 +210,62 @@ public static MethodInfo ReversePatch(MethodBase original, HarmonyMethod standin
 			return PatchFunctions.ReversePatch(standin, original, transpiler, null);
 		}
 
+		/// <summary>Unpatches all methods that were patched by the specified <paramref name="harmonyID"/>. Unpatching is done by repatching methods without patches of this instance.</summary>
+		/// <param name="harmonyID">The Harmony ID to restrict unpatching to a specific Harmony instance.</param>
+		public static void UnpatchID(string harmonyID)
+		{
+			if (string.IsNullOrEmpty(harmonyID))
```
Comment (nike4613) [inline Harmony/Public/Harmony.cs:217]: Since this is a new method altogether, does it not make sense to throw an `ArgumentNullException` in this case? `null` isn't a valid argument value for `harmonyID` now.

Comment (ErisApps) [reply Harmony/Public/Harmony.cs:217]: I don't mind changing it to throw an `ArgumentNullException`, but I just return and log [as per @ghorsington's suggestion in PR #40 ](https://github.com/BepInEx/HarmonyX/pull/40#issuecomment-997419874). Granted, that was regarding the `UnpatchAll` method.

File: Harmony/Public/Harmony.cs [hunk 2]
```diff
@@ -210,52 +210,62 @@ public static MethodInfo ReversePatch(MethodBase original, HarmonyMethod standin
...
+		/// <summary>Unpatches methods by patching them with zero patches. Fully unpatching is not supported. Be careful, unpatching is global</summary>
+		/// <param name="harmonyID">The Harmony ID to restrict unpatching to a specific Harmony instance. Whether this parameter is actually optional is determined by the <see cref="HarmonyGlobalSettings.DisallowGlobalUnpatchAll"/> global flag</param>
+		/// <remarks>When <see cref="HarmonyGlobalSettings.DisallowGlobalUnpatchAll"/> is set to true, the execution of this method will be skipped when no <paramref name="harmonyID"/> is specified.</remarks>
 		///
-		public void UnpatchSelf()
+		[Obsolete("Use UnpatchSelf() to unpatch the current instance", true)]
+		public void UnpatchAll(string harmonyID = null)
 		{
-			UnpatchAll(Id);
+			if (string.IsNullOrEmpty(harmonyID))
```
Comment (ErisApps) [inline Harmony/Public/Harmony.cs:261]: I'm actually not sure whether I should keep checking using `string.IsNullOrEmpty(harmonyID)` or simply check for `harmonyID == null` solely. Emtpy harmonyIDs are technically disallowed because of the check in the ctor for new Harmony instances, so I'm not exactly sure whether I should consider empty ones to be equal to `null` ones

File: Harmony/Public/Harmony.cs [hunk 3]
```diff
@@ -210,52 +210,62 @@ public static MethodInfo ReversePatch(MethodBase original, HarmonyMethod standin
...
+			PatchFunctions.UnpatchConditional(_ => true);
 		}
 
-		/// <summary>Unpatches all methods that were patched by this Harmony instance's ID. Unpatching is done by repatching methods without patches of this instance.</summary>
+		/// <summary>Unpatches methods by patching them with zero patches. Fully unpatching is not supported. Be careful, unpatching is global</summary>
+		/// <param name="harmonyID">The Harmony ID to restrict unpatching to a specific Harmony instance. Whether this parameter is actually optional is determined by the <see cref="HarmonyGlobalSettings.DisallowGlobalUnpatchAll"/> global flag</param>
+		/// <remarks>When <see cref="HarmonyGlobalSettings.DisallowGlobalUnpatchAll"/> is set to true, the execution of this method will be skipped when no <paramref name="harmonyID"/> is specified.</remarks>
 		///
-		public void UnpatchSelf()
+		[Obsolete("Use UnpatchSelf() to unpatch the current instance", true)]
```
Comment (bbepis) [inline Harmony/Public/Harmony.cs:258]: There should be an additional message indicating that the original functionality has been moved to the new static methods

File: Harmony/Public/Harmony.cs [hunk 4]
```diff
@@ -210,52 +210,62 @@ public static MethodInfo ReversePatch(MethodBase original, HarmonyMethod standin
...
 		/// <summary>Unpatches methods by patching them with zero patches. Fully unpatching is not supported. Be careful, unpatching is global</summary>
-		/// <param name="harmonyID">The Harmony ID to restrict unpatching to a specific Harmony instance. Whether this parameter is actually optional is determined by the <see cref="HarmonyGlobalSettings.DisallowGlobalUnpatchAll"/> global flag</param>
-		/// <remarks>This method could be static if it wasn't for the fact that unpatching creates a new replacement method that contains your harmony ID. When <see cref="HarmonyGlobalSettings.DisallowGlobalUnpatchAll"/> is set to true, the execution of this method will be skipped.</remarks>
+		/// <remarks>When <see cref="HarmonyGlobalSettings.DisallowGlobalUnpatchAll"/> is set to true, the execution of this method will be skipped.</remarks>
 		///
-		public void UnpatchAll(string harmonyID = null)
+		public static void UnpatchAll()
 		{
-			if (harmonyID == null)
+			if (HarmonyGlobalSettings.DisallowGlobalUnpatchAll)
```
Comment (bbepis) [inline Harmony/Public/Harmony.cs:239]: Is this setting still necessary if there's new static methods to properly indicate what they do?

**Review by bbepis — CHANGES_REQUESTED** [review]

File: Harmony/Public/Harmony.cs (hunk 1 above)
Comment (bbepis) [reply Harmony/Public/Harmony.cs:217]: +1 for the exception, but the case for an empty string should also be included. Anything that's an invalid ID should not be allowed to continue

File: Harmony/Public/Harmony.cs (hunk 2 above)
Comment (bbepis) [reply Harmony/Public/Harmony.cs:261]: Keep it as `IsNullOrEmpty`. If an empty string gets passed into the second branch, it's going to break things

If you're so inclined, you could make it throw an `ArgumentException` because it wasn't valid input that the method accepted (i.e. it only accepts `null` or a valid harmony ID)

File: Harmony/Public/Harmony.cs (hunk 4 above)
Comment (ErisApps) [reply Harmony/Public/Harmony.cs:239]: Preferably yes, in our case, we want to disallow it altogether because there's not a single good reason why it should be allowed in the first place. This would act as a safeguard. 🙂

File: Harmony/Public/Harmony.cs (hunk 3 above)
Comment (ErisApps) [reply Harmony/Public/Harmony.cs]: Changed.

File: Harmony/Public/Harmony.cs (hunk 4 above)
Comment (ErisApps) [reply Harmony/Public/Harmony.cs:239]: what is not needed though, is the duplicate if check. Will remove the duplicate one.

Comment (bbepis) [reply Harmony/Public/Harmony.cs:239]: Mhhh, I think it's a bit gross but if Horse is fine with it then sure

File: Harmony/Public/Harmony.cs (hunk 2 above)
Comment (ErisApps) [reply Harmony/Public/Harmony.cs:258]: I honestly do indeed feel more inclined to go the exception route for this given that empty strings are invalid anyways.
But then again, I feel like I'd end up in the same discussion as the one in PR #40 so given that this is would be considered a legacy method, I'd rather just go with logging an returning when an empty Id is detected.

File: Harmony/Public/Harmony.cs (hunk 1 above)
Comment (ErisApps) [reply Harmony/Public/Harmony.cs:219]: Changed to throw an `ArgumentNullException` on both `null` and empty `harmonyID`'s.

File: Harmony/Public/Harmony.cs (hunk 4 above)
Comment (ghorsington) [reply Harmony/Public/Harmony.cs:239]: I agree with @bbepis that preventing the call is not necessary here. The main issue of old `UnpatchAll` being called without understanding the implications, as you mentioned in #40, is now resolved by separating the original method into two and forcing users to use `UnpatchSelf` via an obsoletion flag. For base Harmony API, preventing new `Harmony.UnpatchAll` from actually doing what it intends to do is unreasonable, in my opinion.

That said, the old obsoleted `UnpatchAll` should still have this check because of the possibility of confusion, as mentioned in #40.

Suppose you need to prevent the new `Harmony.UnpatchAll` from actually unpatching all (despite the new method existing solely for this purpose). In that case, you can try Harmony patching `Harmony.UnpatchAll` in your mod loader (or even apply a MonoMod Detour) that emits the warning/exception. It might sound a bit hackish, but I think the editing of the *desired behaviour* of a method should be left to the discretion of the mod loader developers.

File: HarmonyTests/Patching/FinalizerPatches.cs [hunk 5]
```diff
@@ -216,7 +216,7 @@ public static void Patch()
 			Assert.NotNull(finalizer, nameof(finalizer));
 
 			var instance = new Harmony("finalizer-test");
-			instance.UnpatchAll("finalizer-test");
+			Harmony.UnpatchID("finalizer-test");
```
Comment (ghorsington) [inline HarmonyTests/Patching/FinalizerPatches.cs:219]:
```suggestion
			instance.UnpatchSelf();
```

**Review by ghorsington — CHANGES_REQUESTED** [review]

File: Harmony/Public/Harmony.cs (hunk 4 above)
Comment (ErisApps) [reply Harmony/Public/Harmony.cs:239]: Gonna give my train of thought on this, but feel free to agree or disagree with anything I'm about to state 🙂 

1. The old instance `UnpatchAll` method is at this point basically just a proxy to the new static method, so the logic for it is contained in a single place.

Surely we could argue that the logic can be moved to only affect the old instance method, but that leads me seamlessly to point 2.

2. Moving said logic would result in inconsistent behavior between the old instance method and the newly created static one. As a user,
3. The infrastructure is already in-place. The global flag was created with the intention to disallow the global UnpatchAll behavior altogether so why not use for the new static method as well. It's opt-in by default so it doesn't affect those who aren't in the known _yet_ nor would it introduce any unexpected behavior.
4. Arguably a moot point, but while the odds of it happening have most likely lowered, it still has a non-zero chance. We were already relying on a harmony patch that patched the instance method to prevent everything from burning down to a crisp. But given that it basically is the same as you already mentioned, it being "hackish", we felt that allowing it to be controlled by a global flag would be the better option. Other than that, there are surely other modding communities who've had the same issue as us, so given them to option to disallow it by simply setting a boolean to `true` only seems beneficial to me.

Long story short, I feel like it wouldn't hurt to have such a safeguard built-in.

Comment (ghorsington) [reply Harmony/Public/Harmony.cs:239]: > The old instance UnpatchAll method is at this point basically just a proxy to the new static method, so the logic for it is contained in a single place.

This is correct for both the point and your argument about moving the argument handling logic to old `UnpatchAll`.

> Moving said logic would result in inconsistent behavior between the old instance method and the newly created static one.

Quoting #40:

> Reasoning for introducing this flag is that, in certain contexts, it is often undesirable to unpatch every single patch at once
> For example, in the case of modloaders where mods can be enabled/disabled at runtime and where accidentally calling the .UnpatchAll() method without an id, could break a whole bunch of other mods.

As the PR implies and as I agreed, the original `UnpatchAll` needed the special argument handler because that method's behaviour was inconsistent in the first place (or, somewhat, its purpose was misinterpreted, which caused inconsistent behaviour). With this PR, the new `UnpatchAll` should behave consistently by always unpatching all instances. If anything, adding special handling as a flag makes the new static method's behaviour inconsistent. In that sense, it would be better called `UnpatchAllExceptAGlobalFlagIsSet`.

In other words, I view #40 change as a necessity stemming from actual documented wrong uses of the method. With the new static method, flag handling is not necessary because this PR splits the behaviour of the old `UnpatchAll` into two.

> The infrastructure is already in-place. The global flag was created with the intention to disallow the global UnpatchAll behavior altogether so why not use for the new static method as well.

If I refer back to the original intention of #40, the PR added the global flag to prevent the improper use of the ambiguously defined `UnpatchAll`.

> It's opt-in by default so it doesn't affect those who aren't in the known yet nor would it introduce any unexpected behavior.

I daresay it will confuse developers who want to unpatch everything for a reason (e.g. temporary profiling). Again, the new static parameterless `UnpatchAll()` should do as it describes. There is no double meaning behind a parameter simply because there is none in the method.
As I see it, Harmony's attempt to merge both use cases (unpatching single instances and unpatching all instances) into a single method is a failure on the API design part. The only reason I accepted #40 was to disable one of the use cases often caused by *misinterpreting the argument's purpose*.

> Arguably a moot point, but while the odds of it happening have most likely lowered, it still has a non-zero chance.

I agree the argument is a tad unwarranted. Notably, you don't need to call `UnpatchAll` to unpatch everything as you can manually construct a `PatchProcessor` and call its own `Unpatch`. I don't see how making the new static `UnpatchAll` having the flag check will prevent unpatching everything. At this point, it seems like clutter like @bbepis suggested.

While I did mention that patching Harmony might seem hackish, manipulating the runtime and libraries in the mod loaders is common. Even in BepInEx, we edit the runtime and the core libraries in various ways to fit the needs of BepInEx. I don't think we should have HarmonyX have all these flags just for the sake of potential use case, but rather to fix or change Harmony behaviour that is a general issue.

***

**To summarise my view:**

* Is using old `UnpatchAll` wrongly an issue? **Yes!** I have personally seen this issue in countless plugins and scripts!
* Does this mean that users will keep using new `UnpatchAll` wrongly to warrant preventing unpatching all instances? **Debatable, but I think no.** This is the policy of the mod loader, no HarmonyX. I'll lean on giving mod developers the freedom and support it with clean API rather than count them for idiots who will go out of their way to write wrong code. It's a difficult thing to balance.

As such, I still stand by that the *old instance* `UnpatchAll` should have the flag check while *new static* `UnpatchAll` should not. If the issue of the flag exists, I suggest renaming the flag to something like `DisallowLegacyUnpatchAll` or `DisallowInstanceUnpatchAll`.

Comment (ErisApps) [reply Harmony/Public/Harmony.cs]: > - Is using old `UnpatchAll` wrongly an issue? **Yes!** I have personally seen this issue in countless plugins and scripts!
> - Does this mean that users will keep using new `UnpatchAll` wrongly to warrant preventing unpatching all instances? **Debatable, but I think no.** This is the policy of the mod loader, no HarmonyX. I'll lean on giving mod developers the freedom and support it with clean API rather than count them for idiots who will go out of their way to write wrong code. It's a difficult thing to balance.

I **definitely** do agree that this is an extremely tedious thing to balance and agree with various arguments that were mentioned, I simply tried to to convince using my point of views, but alas.
Anyways, the things you mentioned, both the renaming of the flag and the moving of the check to the legacy method, as well as the docs of the affected methods and flag have been updated accordingly. Other than that, I also updated the documentation of the new `Harmony.UnpatchAll()` method to make it primarily clear that the unpatching is global.

If there are any other remarks feel free to let me know, but if not, I would kindly like to ask you to make a version available that contains the changes of the past 2 days. 🙂

**Review by ghorsington — APPROVED** [review]

**Review by ManlyMarco — APPROVED** [review]

**Review by bbepis — APPROVED** [review]
